### PR TITLE
MDEV-24377: Accept comma separated addresses as --bind-address value

### PR DESCRIPTION
**Pull request by Paragoumba**
- [x] *The Jira issue number for this PR is: MDEV-24377*

## Description
Allow the option `--bind-address` to take multiple addresses separated by commas as a value. The server then listens on all the resolved addresses.

## How can this PR be tested?

This PR is not tested yet. It can be tested by ensuring it doesn't break existing tests and by adding tests starting the server with multiple valid and invalid values for the `--bind-address` option.

## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

## Backward compatibility
This feature is backward compatible as all the values (single addresses without comma) accepted before implementing this feature are still accepted.

### CODE REVIEW

**Comment by LinuxJedi** [comment]
Hi @Paragoumba, could you please rebase this off 10.11 and add some test cases for this? We should then be able to review it for you.

**Comment by LinuxJedi** [comment]
> Hi @LinuxJedi, sorry for the late response. As you asked I rebased the branch on the 10.11 branch. As fr the test cases I looked into it bu might need some help

Not a problem, I was on vacation until today too :)

For the test cases I suggest looking at `mysql-test/main/bind_address_resolution.*`, copying that and maybe use two addresses in the 127.* range or 127.0.0.1 and 0.0.0.0. Whilst this won't test everything it will at least test for a regression in the parsing and setup of the bind.

**Comment by Paragoumba** [comment]
I created a unit test listening on both localhost (which should be resolved to 127.0.0.1 and ::1) and 192.168.0.1. The test doesn't pass as wsrep can't parse the argument. I tried to parse it ahead of time in `sys_vars.cc` with on ON_CHECK or ON_UPDATE function but couldn't make it work.

**Comment by Paragoumba** [comment]
@LinuxJedi I am not sure how to make the test pass, how should I handle wsrep ?

**Comment by LinuxJedi** [comment]
> @LinuxJedi I am not sure how to make the test pass, how should I handle wsrep ?

Hi!
I have a meeting with the core MariaDB engineering team tomorrow. I'll bring this up and see if I can get some help for you.

**Comment by Paragoumba** [comment]
@LinuxJedi I modified wsrep so it only uses the first address. It shouldn't break anything and refactoring it to use multiple addresses seems to be a task too big for me. All the tests are passing now, including the one I added

**Comment by Paragoumba** [comment]
@LinuxJedi I just reviewed the test I added and it appears it isn't testing properly my changes. A problem I am facing in order to refactor the test is that I can't add the addresses I want to the network interface, so the test fails if I use addresses that localhost doesn't resolve to by default.

**Comment by Paragoumba** [comment]
bump @LinuxJedi

**Review by LinuxJedi — CHANGES_REQUESTED** [review]
I think it is almost there. A couple of minor things. Also can you please add a test that does something like `SHOW VARIABLES LIKE 'bind_address'` to show that it has been processed at that side?

File: sql/mysqld.cc [hunk 1]
```diff
@@ -2291,7 +2292,7 @@ static void activate_tcp_port(uint port,
                               Dynamic_array<MYSQL_SOCKET> *sockets,
                               bool is_extra_port= false)
 {
-  struct addrinfo *ai, *a;
+  struct addrinfo *ai, *a, *b = NULL, *head = NULL;
```
[the remark this answers is not in the document]
Comment (Paragoumba) [reply sql/mysqld.cc]: You are right, I deleted `b` which serves the same purpose as `a`. I believe requesting `head`'s deletion was a mistake on your part and kept it

File: sql/mysqld.cc [hunk 2]
```diff
@@ -2312,16 +2313,50 @@ static void activate_tcp_port(uint port,
...
+      if (b)
+      {
+        b->ai_next= ai;
+      }
+      b= ai;
+      while (b->ai_next)
+      {
+        b= b->ai_next;
+      }
+      if (unlikely(error != 0))
```
[the remark this answers is not in the document]
Comment (Paragoumba) [reply sql/mysqld.cc:2329]: Good catch, I move the error checking

**Comment by grooverdan** [comment]
@vaintroub, the test cases here add a non-existing (at least in some environments) ip of 192.168.0.1 into the bind address tests. Linux obviously tolerates the address while its a 10049 error on Windows. With this it will be a [fatal error](https://buildbot.mariadb.org/#/builders/234/builds/12985/steps/9/logs/stdio). Is the last commit here that aims to make this test pass and only be fatal if no Windows addresses can be bound to acceptable?

**Comment by grooverdan** [comment]
@brad0 do you know what the BSD behaviours are for binding to non-existent addresses?

**Comment by brad0** [comment]
> @brad0 do you know what the BSD behaviours are for binding to non-existent addresses?

I have this feeling like this is something I have seen added over the years, but I'll have to speak to others and get back to you.

File: sql/wsrep_utils.cc [hunk 3]
```diff
@@ -457,7 +457,13 @@ unsigned int wsrep_check_ip (const char* const addr, bool *is_ipv6)
 
   *is_ipv6= false;
 
-  int gai_ret= getaddrinfo(addr, NULL, &hints, &res);
+  char *end;
+  char address[FN_REFLEN];
```
Comment (grooverdan) [inline sql/wsrep_utils.cc:461]: FN_REFLEN was for path lengths so its quite overkill for an address length. `INET6_ADDRSTRLEN` is a good replacement.

File: sql/mysqld.cc [hunk 4]
```diff
@@ -135,6 +135,7 @@
 
 #define mysqld_charset &my_charset_latin1
 
+#define DELIM ','
```
Comment (grooverdan) [inline sql/mysqld.cc:138]: seems used only once, so seems overkill to define it.

**Review by grooverdan — APPROVED** [review]
Overall pretty happy. Just waiting on @vaintroub to approve/reject the Windows handling of bind failures. The other small nits can be cleaned up in the final rebase/squash of commits.

Seems [FreeBSD](http://buildbot.askmonty.org/buildbot/builders/kvm-ports-freebsd130-amd64/builds/3983/steps/mtr/logs/stdio) doesn't like non-existent bind IPs either:
```
2022-10-27  3:25:51 0 [ERROR] Can't start server: Bind on TCP/IP port. Got error: 49: Can't assign requested address
2022-10-27  3:25:51 0 [ERROR] Do you already have another server running on port: 16040 ?
2022-10-27  3:25:51 0 [ERROR] Aborting
````
Errno 49 / [EADDRNOTAVAIL](https://www.unix.com/man-page/FreeBSD/2/errno/)

So just about to push these fixes too.

**Comment by brad0** [comment]
FreeBSD uses IP_BINDANY.


OpenBSD uses SO_BINDANY.

SO_BINDANY allows the socket to be bound to addresses which are not local to the machine, so it can be used to make a transparent proxy. Note that this option is limited to the superuser. In order to receive packets for these addresses, SO_BINDANY needs to be combined with matching outgoing [pf(4)](https://man.openbsd.org/pf.4) rules with the divert-reply parameter. For example, with the following rule the socket receives packets for 192.168.0.10 even if it is not a local address:

```pass out inet from 192.168.0.10 divert-reply```

**Review by LinuxJedi — APPROVED** [review]

**Comment by Paragoumba** [comment]
I am very glad this has been merged. Thank you all for your help, especially @LinuxJedi and @grooverdan :grin:

**Comment by grooverdan** [comment]
Thanks for bearing with a long contribution process. We're very happy to get your code and desired feature into MariaDB server for the use by everyone.

**Pull request merged**